File: Sources/ExcalidrawMath/Angle.swift
import Foundation

/// Angle helpers (`packages/math/src/angle.ts`). Angles are radians unless the
/// name says otherwise; one full turn is `2π`.
public enum Angle {
    /// Wrap an angle into `[0, 2π)` (`normalizeRadians`).
    public static func normalizeRadians(_ angle: Double) -> Double {
        let twoPi = 2 * Double.pi
        return angle < 0 ? (angle.truncatingRemainder(dividingBy: twoPi)) + twoPi
            : angle.truncatingRemainder(dividingBy: twoPi)
    }

    public static func degreesToRadians(_ degrees: Double) -> Double {
        degrees * .pi / 180
    }

    public static func radiansToDegrees(_ radians: Double) -> Double {
        radians * 180 / .pi
    }

    /// Polar coordinates `(radius, angle)` of a cartesian point about the origin
    /// (`cartesian2Polar`).
    public static func cartesianToPolar(_ point: Point) -> (radius: Double, angle: Double) {
        (hypot(point.x, point.y), normalizeRadians(atan2(point.y, point.x)))
    }

    /// Whether `radians` is (close to) a right angle (`isRightAngleRads`).
    public static func isRightAngle(_ radians: Double) -> Bool {
        abs(sin(2 * radians)) < ExcalidrawMath.precision
    }

    /// Whether `angle` lies within `[min, max]`, accounting for wrap-around
    /// (`radiansBetweenAngles`).
    public static func radiansBetween(_ angle: Double, min: Double, max: Double) -> Bool {
        let a = normalizeRadians(angle)
        let lo = normalizeRadians(min)
        let hi = normalizeRadians(max)
        if lo < hi { return a >= lo && a <= hi }
        return a >= lo || a <= hi // range wraps past 0
    }

    /// Smallest absolute difference between two angles (`radiansDifference`).
    public static func radiansDifference(_ a: Double, _ b: Double) -> Double {
        var diff = normalizeRadians(a) - normalizeRadians(b)
        if diff < -Double.pi {
            diff += 2 * .pi
        } else if diff > Double.pi {
            diff -= 2 * .pi
        }
        return abs(diff)
    }
}

File: web/packages/excalidraw-svelte/src/math/angle.ts
import { PRECISION, type Point } from "./point.js";

const TWO_PI = 2 * Math.PI;

/** Wrap an angle into `[0, 2π)` (`normalizeRadians`). */
export function normalizeRadians(angle: number): number {
  const remainder = angle % TWO_PI;
  return angle < 0 ? remainder + TWO_PI : remainder;
}

export function degreesToRadians(degrees: number): number {
  return (degrees * Math.PI) / 180;
}

export function radiansToDegrees(radians: number): number {
  return (radians * 180) / Math.PI;
}

/** Polar `(radius, angle)` of a cartesian point about the origin (`cartesian2Polar`). */
export function cartesianToPolar(point: Point): { radius: number; angle: number } {
  return {
    radius: Math.hypot(point.x, point.y),
    angle: normalizeRadians(Math.atan2(point.y, point.x)),
  };
}

/** Whether `radians` is (close to) a right angle (`isRightAngleRads`). */
export function isRightAngle(radians: number): boolean {
  return Math.abs(Math.sin(2 * radians)) < PRECISION;
}

/** Whether `angle` lies within `[min, max]`, accounting for wrap-around. */
export function radiansBetween(angle: number, min: number, max: number): boolean {
  const a = normalizeRadians(angle);
  const lo = normalizeRadians(min);
  const hi = normalizeRadians(max);
  if (lo < hi) return a >= lo && a <= hi;
  return a >= lo || a <= hi; // range wraps past 0
}

/** Smallest absolute difference between two angles (`radiansDifference`). */
export function radiansDifference(a: number, b: number): number {
  let diff = normalizeRadians(a) - normalizeRadians(b);
  if (diff < -Math.PI) {
    diff += TWO_PI;
  } else if (diff > Math.PI) {
    diff -= TWO_PI;
  }
  return Math.abs(diff);
}

/** Namespace object mirroring the Swift `Angle` enum. */
export const Angle = {
  normalizeRadians,
  degreesToRadians,
  radiansToDegrees,
  cartesianToPolar,
  isRightAngle,
  radiansBetween,
  radiansDifference,
} as const;
